korangar/src/loaders/animation/mod.rs

```diff
@@ -0,0 +1,474 @@
+            size: Vector2::new(new_width, new_height),
+            upleft: Vector2::zero(),
+            offset: Vector2::new(upleft_x + (new_width - 1) / 2, upleft_y + (new_height - 1) / 2),
+            remove_offset: Vector2::zero(),
+            frameparts: new_frameparts,
+        }
+    }
+}
+
+#[derive(Clone, PrototypeElement)]
+pub struct Animation {
+    #[hidden_element]
+    pub frames: Vec<Frame>,
+}
+
+#[derive(Clone, PrototypeElement)]
+pub struct AnimationPair {
+    pub sprites: Arc<Sprite>,
+    pub actions: Arc<Actions>,
+}
+
+#[derive(Clone, PrototypeElement)]
+pub struct AnimationData {
+    pub animation_pair: Vec<AnimationPair>,
+    pub animations: Vec<Animation>,
+    pub delays: Vec<f32>,
+    #[hidden_element]
+    pub entity_type: EntityType,
+}
+
+// This function generates the convertion to the big rectangle
+fn convert_coordinate(coordinate: Vector2<i32>, size: Vector2<i32>) -> Vector2<f32> {
+    let x = (coordinate.x as f32 / size.x as f32 - 0.5) * 2.0;
+    let y = 2.0 - (coordinate.y as f32 / size.y as f32) * 2.0;
+    return Vector2::<f32>::new(x, y);
+}
+
+impl AnimationData {
+    pub fn render(
+        &self,
+        animation_state: &AnimationState,
+        camera_direction: usize,
+        head_direction: usize,
+    ) -> Vec<(Arc<Texture>, Vec<Vector2<f32>>, Vector2<f32>, Vector2<i32>, f32, Color, bool)> {
+        let direction = (camera_direction + head_direction) % 8;
+        let aa = animation_state.action * 8 + direction;
+        let delay = self.delays[aa % self.delays.len()];
+        let animation = &self.animations[aa % self.animations.len()];
+
+        let factor = animation_state
+            .factor
+            .map(|factor| delay * (factor / 5.0))
+            .unwrap_or_else(|| delay * 50.0);
+
+        let frame_time = animation_state
+            .duration
+            .map(|duration| animation_state.time * animation.frames.len() as u32 / duration)
+            .unwrap_or_else(|| (animation_state.time as f32 / factor) as u32);
+        // TODO: work out how to avoid losing digits when casting timg to an f32. When
+        // fixed remove set_start_time in MouseCursor.
+
+        let mut frame = &animation.frames[0];
+        let size = frame.size;
+        let time = frame_time as usize % animation.frames.len();
+
+        // Remove Doridori animation from Player
+        if self.entity_type == EntityType::Player && animation_state.action == 0 {
+            frame = &animation.frames[0];
+        } else {
+            frame = &animation.frames[time];
+        }
+
+        let mut datas = Vec::new();
+        for framepart in frame.frameparts.iter() {
+            let texture;
+            let position;
+            let animation_index = framepart.animation_index as usize;
+            let sprite_number = framepart.sprite_number as usize;
+            texture = &self.animation_pair[animation_index].sprites.textures[sprite_number];
+            position = Vector2::new(

```

Comment:
It would be great if you could explain the operation here. Especially the use "/2" and "/10" parts.

Reply:
Explained in a better way